### ISSUE TITLE
Cloudflare issues with token in config

### ISSUE THREAD

**Issue by lilHermit** [pulumi/pulumi-cloudflare#280]
### What happened?

I'm having an issue with `cloudflare:apiToken` stored as a secret where it looks like it's being ignore however exporting the variable or passing into the call both work.

### Steps to reproduce

Peforming 
``` 
AWS_PROFILE=my_profile pulumi config set cloudflare:apiToken --secret rEdAcTeD
AWS_PROFILE=my_profile pulumi refresh
```

results in all my cloudflare resources errorring like

```
    error: Preview failed: could not validate provider configuration: 1 error occurred:
    	* Invalid combination of arguments: "api_key": one of `api_key,api_token` must be specified
```

If I provide the token via the command line like so, works
```
AWS_PROFILE=my_profile CLOUDFLARE_API_TOKEN=rEdAcTeD pulumi refresh
```
or
```
export CLOUDFLARE_API_TOKEN=rEdAcTeD
AWS_PROFILE=my_profile pulumi refresh
```

### Expected Behavior

Refresh without errors

### Actual Behavior

Errors for all cloudflare resources

### Output of `pulumi about`

``` 
CLI          
Version      3.38.0
Go Version   go1.19
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.11.0
cloudflare  4.9.0
docker      3.4.1
nodejs      unknown
random      4.8.2
terraform   5.6.6

Host     
OS       debian
Version  11.4
Arch     x86_64

This project is written in nodejs: executable='/usr/local/bin/node' version='v16.16.0'

Current Stack: global

--Removed 280 resources

Found no pending operations associated with global

Backend        
Name           c94133a78fad
URL            s3://my-s3-state/path
User           root
Organizations  

Dependencies:
NAME                VERSION
@pulumi/awsx        0.40.0
@pulumi/pulumi      3.38.0
@types/lodash       4.14.184
@types/node         14.18.26
lodash              4.17.21
@pulumi/aws         5.11.0
@pulumi/cloudflare  4.9.0
@pulumi/random      4.8.2
@pulumi/terraform   5.6.6

### Additional context

I've tried with the latest cloudflare and pulumi versions too

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by viveklak** [comment]
My understanding is that this provider doesn't really support the config option even though the installation guide says so here: https://www.pulumi.com/registry/packages/cloudflare/installation-configuration/#configuring-the-provider
@guineveresaenger sounds like with the addition of https://github.com/pulumi/pulumi-terraform-bridge/pull/606/ we should be able to make this sort of support a lot simpler to add?

**Comment by viveklak** [comment]
To be precise there are two options - fix the documentation to remove the config option setting or add support for config options. The latter would be more desirable.

**Comment by viveklak** [comment]
@stack72 to make sure my understanding is correct here.

**Comment by guineveresaenger** [comment]
https://github.com/pulumi/pulumi-terraform-bridge/pull/606 is really only a set of helper functions for warning - they won't affect configuration options at all. Let me look at what is happening here - my guess is that there's something going wring with the `secret` part.

@lilHermit what happens when you use `pulumi config set cloudflare:apiToken <redacted>` - but _without_ the `--secret` flag?

**Comment by viveklak** [comment]
Thanks @guineveresaenger - I had misunderstood a couple of things. 

Guin and I looked through this together and I think we understand what is happening here.
@lilHermit  - it sounds like you have an existing cloudflare provider initialized using the api key. Can you post some details on how you are passing that to the provider? i.e. is through the config parameters, passed explicitly in the constructor or the environment variables?

[Only one of api key and api token](https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs#api_key) can be passed to the provider as specified but this is missing from the Pulumi provider's documentation. When you specify the api token through pulumi's config parameter, the terraform provider's validation is triggered and it fails with the error you saw. However, the validation is skipped when the token is specified through the environment variable. Moreover, the TF provider prefers the api token over the api key when specified using the environment variables. Overall this is expected behavior though I admit the documentation does not make this obvious which we will fix.

In summary, you want to only set one of the API key and API tokens as the error suggests.

**Comment by lilHermit** [comment]
> @lilHermit what happens when you use `pulumi config set cloudflare:apiToken <redacted>` - but _without_ the `--secret` flag?

It errors because it believes it's a secret
```
error: config value for 'cloudflare:apiToken' looks like a secret; rerun with --secret to encrypt it, or --plaintext if you meant to store in plaintext```

**Comment by lilHermit** [comment]
> Thanks @guineveresaenger - I had misunderstood a couple of things.
> 
> Guin and I looked through this together and I think we understand what is happening here. @lilHermit - it sounds like you have an existing cloudflare provider initialized using the api key. Can you post some details on how you are passing that to the provider? i.e. is through the config parameters, passed explicitly in the constructor or the environment variables?
> 
> [Only one of api key and api token](https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs#api_key) can be passed to the provider as specified but this is missing from the Pulumi provider's documentation. When you specify the api token through pulumi's config parameter, the terraform provider's validation is triggered and it fails with the error you saw. However, the validation is skipped when the token is specified through the environment variable. Moreover, the TF provider prefers the api token over the api key when specified using the environment variables. Overall this is expected behavior though I admit the documentation does not make this obvious which we will fix.
> 
> In summary, you want to only set one of the API key and API tokens as the error suggests.
I am not initialising the provider manually I'm letting the code do this as per the docs. It did work for sometime and then all of a sudden stopped. I'm only using API token not key. As I said using the ENV far the provider picks up the apiToken but not from the config. Whilst ENV works ok for local runs it's not something I want to do in CICD.

**Comment by guineveresaenger** [comment]
hi @lilHermit - here are a few ways to check your configuration settings to make sure you're not accidentally providing both `api_key` and `api_token`.

- check in your `Pulumi.<stackname>.yaml` to see if `api_key` is hiding in there
- run `pulumi config get `cloudflare:api_key` and `pulumi config get `cloudflare:api_token`
- check on pulumi.com on `youraccount/yourproject/yourstack` under "Configuration" for any discrepancies
- You can also look for `api_key` and `api_token` in the output of `pulumi stack export` - it is possible you may have set them on individual resources, although that is really unlikely to be possible for these values - I'm not super familiar with Cloudflare though.

If the error message is correct (and it might not be; these OnlyOneOf settings have some interesting logic sometimes) then you have an `api_key` sitting around quietly somewhere.

At this point, we'd need to see a code sample to verify we can repro this behavior and look into it. 🙇‍♀️

**Comment by lilHermit** [comment]
> hi @lilHermit - here are a few ways to check your configuration settings to make sure you're not accidentally providing both `api_key` and `api_token`.
Ok strangely the documentation says `apiKey` and `apiToken` [Cloudflare docs](https://www.pulumi.com/registry/packages/cloudflare/installation-configuration/) but the error implies `api_key` / `api_token`. I have changed my `Pulumi.<stackname>.yaml` to `api_token` to test but still the same

> * check in your `Pulumi.<stackname>.yaml` to see if `api_key` is hiding in there

There is an `apiToken` but this gets ignored, the only way to provide it is via `CLOUDFLARE_API_TOKEN` env variable

> * run `pulumi config get cloudflare:api_key`and `pulumi config get cloudflare:api_token`
``` 
root@05eec48ecb07:/pulumi/projects# AWS_PROFILE=my-profile pulumi config get cloudflare:apiKey
error: configuration key 'cloudflare:apiKey' not found for stack 'global'

root@05eec48ecb07:/pulumi/projects# AWS_PROFILE=my-profile pulumi config get cloudflare:apiToken
<rEdAcTeDtOkEn>
```

> * check on pulumi.com on `youraccount/yourproject/yourstack` under "Configuration" for any discrepancies

We're not using pulumi.com for state (s3 bucket)

> * You can also look for `api_key` and `api_token` in the output of `pulumi stack export` - it is possible you may have set them on individual resources, although that is really unlikely to be possible for these values - I'm not super familiar with Cloudflare though.

```
root@05eec48ecb07:/pulumi/projects# AWS_PROFILE=my-profile pulumi stack export | grep apiToken
root@05eec48ecb07:/pulumi/projects# AWS_PROFILE=my-profile pulumi stack export | grep apiKey  
```

> If the error message is correct (and it might not be; these OnlyOneOf settings have some interesting logic sometimes) then you have an `api_key` sitting around quietly somewhere.
> 
> At this point, we'd need to see a code sample to verify we can repro this behavior and look into it. bowing_woman

I'll create  repo this morning

**Comment by lilHermit** [comment]
Ok on creating a "proof of issue" repo it works fine so there must be something somewhere in the stack/state

In my program
```
AWS_PROFILE=my-profile pulumi stack export | egrep -i "api_key|apiKey|api_token|apiToken"
```
Returns no matches

In the "proof of issue"
```
pulumi stack export | egrep -i "api_key|apiKey|api_token|apiToken"
                    "apiToken": {
                    "apiToken": {
```

I then found [this thread](https://github.com/pulumi/pulumi/issues/4981) 

Running the following fixed the provider and now a `refresh` works without issues
```
AWS_PROFILE=my-profile pulumi up --target "urn:pulumi:global::my-program::pulumi:providers:cloudflare::default_4_10_0" --skip-preview
```

Confirmation of the token now in the state
```
AWS_PROFILE=my-profile pulumi stack export | egrep -i "api_key|apiKey|api_token|apiToken"
                    "apiToken": {
                    "apiToken": {
```

**Summary**
There was neither `aplToken` or `apiKey` in the state so the error message is misleading. I assume the provider had got out of sync with the config, I noticed the `pulumi config refresh` command which removed the `apiToken` from the config it might be worth haveing the opposite to sync state with the config?!

**Comment by guineveresaenger** [comment]
Hi @lilHermit - I am so sorry for all this back and forth. I believe the root cause here is  that you're correct, our documentation is wrong, and I have [pushed a fix for that here](https://github.com/pulumi/registry/pull/1720).

Since it sounds like you've figured out a way forward, I'm going to close this issue once the registry PR merges.

Thank you for reporting and apologies for the misleading documentation.

**Comment by guineveresaenger** [comment]
Hi @lilHermit - apologies for the delay.

I tried this with the most recent version of Pulumi and could not reproduce this issue. Please re-open and ensure to add the "needs-triage" label if you're still having trouble!
